---
title: "Untitled"
author: "Amy Gill"
date: "8/8/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Transcript quantification assessment

```{r}
# exon length
lengths = c(100,200,300,100,100)

# exon pattern in transcripts 1-3
mat = cbind(c(1,1,0,1,0),c(1,1,1,1,1),c(0,1,1,0,1))

# length of transcripts
lengths %*% mat

# 1000 reads
w = 1000
 
# counts per exon
counts = c(125,350,300,125,100)

# guess at theta for maximum likelihood estimate
theta.hat = c(1, 2, 3) / 10000

# result of guess - counts too low and out of proportion
mat %*% theta.hat * lengths * w

# roughly estimate theta by solving system of equations
LHS = counts/(lengths * w)
lm.fit(mat, LHS)$coefficients
```

### Q1

```{r}
theta.hat = c(.00075, .0005, .0005)
mat %*% theta.hat * lengths * w
```

### Q2

```{r}
counts = c(60,320,420,60,140)
LHS = counts/(lengths * w)
lm.fit(mat, LHS)$coefficients
```

## Transcriptome alignment assessment

```{r}
genes <- read.table("SRR1039508.genes.results", header=TRUE)
isoforms <- read.table("SRR1039508.isoforms.results", header=TRUE)
```

```{r}
fpkm.per.gene <- split(isoforms$FPKM, isoforms$gene_id)
head(sapply(fpkm.per.gene, sum))
head(genes$FPKM)
```

```{r}
library(dplyr)
isoforms %>% group_by(gene_id) %>% summarize(sum = sum(FPKM))
```

```{r}
genes2 <- genes[genes$FPKM > 0,]
genes2$gene_id <- droplevels(genes2$gene_id)
isoforms2 <- isoforms[isoforms$gene_id %in% genes2$gene_id,]
isoforms2$gene_id <- droplevels(isoforms2$gene_id)
```

```{r}
stopifnot(all(genes2$gene_id == levels(isoforms2$gene_id)))
```

### Q1

```{r}
median(genes2$FPKM)
```

### Q2

```{r}
max.iso = sapply(split(isoforms2$IsoPct, isoforms2$gene_id), max)
mean(max.iso > 95)

# ...or with dplyr:
isoforms2 %>% group_by(gene_id) %>% summarize(max.iso = max(IsoPct)) %>% summarize(answer = mean(max.iso > 95))
```

### Q3

```{r}
boxplot(split(log10(genes2$FPKM), cut(max.iso, 5)), xlab="max.iso", ylab="log10 FPKM")
```

### Q4

```{r}
num.iso <- as.numeric(table(isoforms2$gene_id))
plot(num.iso, max.iso)
```

```{r}
barplot(table(num.iso))
barplot(table(num.iso[num.iso < 15]))
```

## Install bioc assessment

```{r}
BiocManager::version()
```

```{r}
BiocManager::install()
```

## Count matrix assessment

```{r}
library(pasillaBamSubset)
bam.file <- untreated3_chr4()
library(Rsamtools)
bf <- BamFile(bam.file)
```

